### ISSUE TITLE
Some programs can't find OCR text in Tesseract's PDFs (3.04)

### ISSUE THREAD

**Issue by jbarlow83** [tesseract-ocr/tesseract#182]
While Acrobat XI can find text in a PDF, it appears that [poppler](https://en.wikipedia.org/wiki/Pdftotext)'s `pdftotext` program, OS X's Preview app, and the library [PyPDF2](https://github.com/mstamy2/PyPDF2/tree/master/PyPDF2)'s extractText() function all fail to locate text. It seems that Tesseract is encoding text in a way that makes it inaccessible to many PDF viewers.

`pdftotext` produces empty output.
Preview app allows highlighting of text in the appropriate locations, but it cannot be copied to the clipboard or searched.
PyPDF2 extractText also produces an empty string as text.

**Comment by amitdo** [comment]
See #170.

**Comment by jbarlow83** [comment]
#170 might be related, but the files I checked did not have tilted or skewed text.

Input file:
![linnsequencer](https://cloud.githubusercontent.com/assets/1825843/12100857/6fba9e66-b2e6-11e5-8df1-d928e75a1b44.jpg)

Output:
[linn.pdf](https://github.com/tesseract-ocr/tesseract/files/77727/linn.pdf)

tesseract version
``
tesseract 3.04.00
 leptonica-1.72
  libjpeg 8d : libpng 1.6.19 : libtiff 4.0.6 : zlib 1.2.5
``

**Comment by amitdo** [comment]
Chromium's pdf reader output (cut&paste):

>The LinnSequencer
32 Track MIDI Sequence Recorder
The LinnSequencer is state—of—the-art composition and performance tool for the professional musician. It is
extremely powerful, yet amazingly simple to learn and use. It’s many remarkable features include:
Operation is similar to multi-track tape recorder with PLAY, STOP, RECORD, FAST
FORWARD, REWIND, and LOCATE controls.
Each of the 100 sequences contains 32 simultaneous, polyphonic tracks. Each track may
be assigned to one of 16 MIDI channels. Simultaneously plays up to 16 polyphonic
synthesizers
Ultra-fast 1/2 disk drive stores complex songs in seconds and holds over 110,000 notes
per disk!
One or all tracks may be TRANSPOSED at the touch of key.
Exclusive real—time ERASE function makes editing FAST.
Exclusive REPEAT function automatically repeats any held notes at pre-selected
rhythmic value.
TIMING CORRECTION works during playback and operates without ‘chopping’ notes.
Optional SMPTE time code synchronization.
Optional remote control.
Recording Sequence
To record sequence, simply press RECORD and PLAY,
then play your MIDI keyboard in time to the Sequencer’s
click track. When the sequence loops back around to bar 1,
you’ll hear what you played—only all timing errors will be
corrected! (Timing correction may be adjusted 0r defeated).
Any additional notes played will be added into the track
—existing notes are not erased while recording!
FAST FORWARD, REWIND, and LOCATE controls
may be used at any time to quickly access any location in
your sequence for spot-recording. To overdub new part,
select different track and start recording—while you
record, the first‘track will play in perfect sync (unless you
MUTE it, or SOLO another track). In this way, up to 32
tracks may be overdubbed! All MIDI effects are recorded
including pitch bend, modulation, velocity, aftertouch,
sustain pedal, and program changes!
Editing
To erase wrong note, simply hold ERASE and press
the note to be erased just before it plays in the sequence-—
when played back, it will be gone. Notes may also be
added, erased, or changed using the SINGLE STEP function.
To overdub notes at specific points within sequence,
Additional Features
simply use LOCATE, FAST FORWARD, or REWIND to
find the desired bar number, then start recording.
The INSERT/COPY function allows you to move bars
from one location to another—in the same sequence or
different one. For example, you might insert copy of the
first verse between the second chorus and the bridge.
DELETE BARS operates the same way to remove
unwanted sections.
Creating Song
One way to create song is to record each track all the
way through (up to 999 bars). Another way is to record
each basic section (verse, chorus, etc.) in individual
sequences, then use the CREATE SONG function to “chain”
them together. CREATE SONG will then automatically
copy all the parts into new sequence. If desired, you can
even set the last few bars to repeat infinitely, for fadeout.
Composition Without Compromise
The technology you use should never be so complex that
it interferes with the creative process. That’s precisely why
the LinnSequencer is designed to let you compose, record
and edit while devoting your undivided attention to your
music. See your Linn dealer today for demonstration!
Simple, easy to learn operation—the 32 character LCD display clearly guides you through all operations. If needed, the
HELP button displays additional explanations.
Non-destructive recording—existing notes are not erased while recording.
Two FOOTSWITCH INPUTS may be assigned to remotely control many of the commonly used functions, including
ERASE, REPEAT, PLAY/STOP, or LOCATE.
Two TRIGGER OUTPUTS may be programmed to output pulses at any selected note value.
Will sync to standard LinnDrum or Linn 9000 sync tone.
Utilizes ultra high—speed, MHZ 80186 16 bit computer internally for FAST operation.
TEMPO may be specified in BEATS-PER—MINUTE or FRAMES-PER—BEAT at 24, 25, or 30 frames per second,
(even drop frame!)
TEMPO may be entered numerically, adjustable in tenths of Beat-Per-Minute increments, or by tapping quarter notes
on the TAP TEMPO button.
TEMPO CHANGES may be programmed into sequence, with smooth transitions if desired.
Any TIME SIGNATURE may be used, and may be changed within song.
EDI]
Linn Electronics, Inc.
18720 Oxnard Street, Tarzana, CA 91356
(818) 708-8131 TELEX #298949 LINN U

**Comment by amitdo** [comment]
I run Tesseract with your jpg image:

> tesseract i182.jpg i182 -l eng txt pdf hocr

Evince (base on poppler) output (cut&paste):

>The LinnSequencer
32 Track MIDI Sequence Recorder
The LinnSequencer is a state—of—the-art composition and performance tool for the professional musician. It is
extremely powerful, yet amazingly simple to learn and use. It’s many remarkable features include:
0 Operation is similar to multi-track tape recorder with PLAY, STOP, RECORD, FAST
FORWARD, REWIND, and LOCATE controls.
0 Each of the 100 sequences contains 32 simultaneous, polyphonic tracks. Each track may
be assigned to one of 16 MIDI channels. Simultaneously plays up to 16 polyphonic
synthesizers !
0 Ultra-fast 3 1/2 ” disk drive stores complex songs in seconds and holds over 110,000 notes
per disk!
0 One or all tracks may be TRANSPOSED at the touch of a key.
0 Exclusive real—time ERASE function makes editing FAST.
0 Exclusive REPEAT function automatically repeats any held notes at a pre-selected
rhythmic value.
0 TIMING CORRECTION works during playback and operates without ‘chopping’ notes.
0 Optional SMPTE time code synchronization.
0 Optional remote control.
Recording a Sequence
simply use LOCATE, FAST FORWARD, or REWIND to
find the desired bar number, then start recording.
The INSERT/COPY function allows you to move bars
from one location to another—in the same sequence or a
different one. For example, you might insert a copy of the
first verse between the second chorus and the bridge.
DELETE BARS operates the same way to remove
unwanted sections.
To record a sequence, simply press RECORD and PLAY,
then play your MIDI keyboard in time to the Sequencer’s
click track. When the sequence loops back around to bar 1,
you’ll hear what you played—only all timing errors will be
corrected! (Timing correction may be adjusted 0r defeated).
Any additional notes played will be added into the track
—existing notes are not erased while recording!
FAST FORWARD, REWIND, and LOCATE controls
may be used at any time to quickly access any location in
your sequence for spot-recording. To overdub a new part,
select a different track and start recording—while you
record, the first‘track will play in perfect sync (unless you
MUTE it, or SOLO another track). In this way, up to 32
tracks may be overdubbed! All MIDI effects are recorded
including pitch bend, modulation, velocity, aftertouch,
sustain pedal, and program changes! One way to create a song is to record each track all the
way through (up to 999 bars). Another way is to record
each basic section (verse, chorus, etc.) in individual
sequences, then use the CREATE SONG function to “chain”
them together. CREATE SONG will then automatically
copy all the parts into a new sequence. If desired, you can
even set the last few bars to repeat infinitely, for a fadeout.
Editing Composition Without Compromise
To erase a wrong note, simply hold ERASE and press
the note to be erased just before it plays in the sequence-—
when played back, it will be gone. Notes may also be
added, erased, or changed using the SINGLE STEP func-
tion. To overdub notes at specific points within a sequence, The technology you use should never be so complex that
it interferes with the creative process. That’s precisely why
the LinnSequencer is designed to let you compose, record
and edit while devoting your undivided attention to your
music. See your Linn dealer today for a demonstration!
Creating a Song
Additional Features
0 Simple, easy to learn operation—the 32 character LCD display clearly guides you through all operations. If needed, the
HELP button displays additional explanations.
0 Non-destructive recording—existing notes are not erased while recording.
0 Two FOOTSWITCH INPUTS may be assigned to remotely control many of the commonly used functions, including
ERASE, REPEAT, PLAY/STOP, or LOCATE.
0 Two TRIGGER OUTPUTS may be programmed to output pulses at any selected note value.
0 Will sync to standard LinnDrum or Linn 9000 sync tone.
0 Utilizes ultra high—speed, 8 MHZ 80186 16 bit computer internally for FAST operation.
0 TEMPO may be specified in BEATS-PER—MINUTE or FRAMES-PER—BEAT at 24, 25, or 30 frames per second,
(even drop frame!)
0 TEMPO may be entered numerically, adjustable in tenths of a Beat-Per-Minute increments, or by tapping quarter notes
on the TAP TEMPO button.
0 TEMPO CHANGES may be programmed into a sequence, with smooth transitions if desired.
0 Any TIME SIGNATURE may be used, and may be changed within a song.
EDI]
Linn Electronics, Inc.
18720 Oxnard Street, Tarzana, CA 91356
(818) 708-8131 TELEX #298949 LINN UR

**Comment by amitdo** [comment]
Here are the output files...

[i182.pdf](https://github.com/tesseract-ocr/tesseract/files/77832/i182.pdf)
[i182.txt](https://github.com/tesseract-ocr/tesseract/files/77833/i182.txt)
[i182-hocr.zip](https://github.com/tesseract-ocr/tesseract/files/77840/i182-hocr.zip)

**Comment by jbarlow83** [comment]
Chrome's PDF reader works for me.

I have poppler 0.39.0 installed (homebrew/OS X/El Capitan).

I believe I found the reason. It appears that the readers that struggle with it do not support Tesseract's usage of hexadecimal code points rather than literal characters in the output stream.

The PostScript content stream for this page as generated by Tesseract for the first word, "The" appears as follows:
```Postscript
 Tz [ <0054><0068><0065> ] TJ  
```

where <0054> = U+0054 = T, <0068> = U+0068 = h, etc. I have run into other situations where this hexadecimal notation introduces errors.

Acrobat generates the equivalent segment with ASCII literals.
```Postscript
[...omitted...] Tm
(The )Tj
```

Longer excerpts for comparison:

Tesseract

```Postscript
BT    
3 Tr 1 0 0 1 211.68 744 Tm /f-0-0 21 Tf 117.334 Tz [ <0054><0068><0065> ] TJ  
```

Acrobat

```Postscript
BT
0.196 0.184 0.188 rg
/T1_0 1 Tf
-0.035 Tc 3 Tr 23.4905 0 0 23.7001 211.43 744.24 Tm
(The )Tj
ET
```

**Comment by amitdo** [comment]
Did you see my latest comments?
The latest commit from the repo produce better pdf results than version 3.04.

**Comment by jbarlow83** [comment]
Yes. Preview and poppler are still incapable of reading your i182.pdf. I observed no difference.

My comparison didn't address how Acrobat handles Unicode and Unicode literals cannot appear in Postscript so I checked how this is done. When Acrobat encodes a Unicode string it uses UTF-16 big endian code points in hexadecimal, like this:
```Postscript
... Tm
<4E8B5F97771F5BF9770B89C152A066F4591A5C11>Tj
```

That string encodes 10 characters all below U+7FFF, which are these:
事得真对看见加更多少

So it appears that Tesseract's method of encoding strings is nonstandard.

**Comment by amitdo** [comment]
My libpopler is 0.24.5. Ubuntu 14.04.

>pdftotext i182.pdf i182t.txt

Here is the pdftotext output:
[i182t.txt](https://github.com/tesseract-ocr/tesseract/files/77872/i182t.txt)

**Comment by amitdo** [comment]
cc: @jbreiden
jbreiden wrote Tesseract's pdf renderer code.

**Comment by jbarlow83** [comment]
Okay, for some reason pdftotext won't output to stdout by will save to a file. Quick guess is that it suppresses its output if high ASCII characters are present, and that file contains some representing n-dashes or smart quotes. Both poppler 0.24.5 and 0.34 behave as expected when asked to save to a file, so the text stream is accessible to pdftotext.

That said, OS X Preview and parsers like PyPDF2 still struggle with how Tesseract encodes text, as far as I can tell. 

I checked that reportlab also encodes text strings in the manner of Acrobat, and Preview has no problems with PDFs produced by Tesseract -> hOCR -> reportlab PDF. This is an example of such a file:

[linn_hocr_unc.pdf](https://github.com/tesseract-ocr/tesseract/files/77885/linn_hocr_unc.pdf)

**Comment by amitdo** [comment]
From https://groups.google.com/forum/?hl=en#!topic/tesseract-dev/XllxjvK5HtU

>Jeff Breidenbach 	7/17/15
PROBLEM #2: PDF
I was looking at a PDF problem report and noticed that Tesseract PDF output
is no longer validating. (It fails qpdf --check). As the author of the pdf module,
I'm biased, but producing corrupt data is a disaster and I think we need to cut
a new release once it is figured out. Most PDF viewers will recover and silently 
ignore, but this is no good at all. I wonder what happened.

**Comment by amitdo** [comment]
Try this to output to stdout:
 
>    pdftotext i182.pdf -

Jeff mentioned qpdf.
Links:
 http://qpdf.sourceforge.net 
https://github.com/qpdf/qpdf

**Comment by jbarlow83** [comment]
Qpdf says it okay, but it doesn't check everything.

**Comment by jbreiden** [comment]
I might not have time to take a look until Wednesday. Validators of various flavors  include jhove, jhove-pdf-a, pdfbox, ITextRUPS, and http://www.pdf-tools.com/pdf/validate-pdfa-online.aspx. I did compatibility testing with Apple's Preview at design time, but I don't test against it regularly. Never tried PyPDF2.  If I had to guess right now, I'd suspect it might be the invisible font improvement that was written for better ghostscript compatibility. Unlikely to be the hex encoding.

**Comment by jbreiden** [comment]
(Note that Tesseract PDFs are *not* expected to be PDF/A compliant)

**Comment by jbreiden** [comment]
Looking at issue 181, it's looking more and more like Preview is unhappy with the revised glyphless font, possible due to the zero advance width. Will try to borrow a Mac and play with it, hopefully on Wednesday.

**Comment by jbarlow83** [comment]
@jbreiden I agree that the glyphless font issue seems more probable.

Aside: I wouldn't trust JHOVE for PDF validation. For JHOVE to approve is better than not approving, but its analysis is rudimentary, and in my experience it produce more false positives and negatives than useful diagnostics.

**Comment by jbreiden** [comment]
I produced this PDF using Tesseract, then borrowed a laptop running Mac OS X version 10.10.5 and was able to both search and copy-paste from Preview (Although the copy-paste highlighting was kind of weird). My testing copy of Tesseract is not completely synchronized with GitHub, so if needed we can investigate that. How does this PDF perform for you on Preview, @jbarlow83 ?

[2.pdf](https://github.com/tesseract-ocr/tesseract/files/79277/2.pdf)

There is also an alternative invisible font here, that contains an advanceWidth. I think it can be swapped in for tessdata/pdf.ttf. It has a side effect of making highlighting look even more bizarre in evince. I don't notice any compatibility differences at all, but putting this up for testing.

https://github.com/behdad/tofudetector/blob/master/tofu.ttf?raw=true

Here's the original image (I was actually using TIFF but GitHub doesn't let me attach that)

![relativity](https://cloud.githubusercontent.com/assets/4961958/12132423/45519c7a-b3d0-11e5-9700-3179d131ae9e.png)

**Comment by jbarlow83** [comment]
Doesn't work in Preview OS X 10.11.2 (highlights properly, but no copy-paste or search). I have access to two other OS X machines - will check those later day.

I check with my iPhone too. Both Chrome iOS (PDFium?) via Gmail app and Safari struggle to highlight text (they only allow highlighting a single character) and cannot copy.

**Comment by jbreiden** [comment]
This one uses the alternate font that has an advance width.

[alternate.pdf](https://github.com/tesseract-ocr/tesseract/files/79359/alternate.pdf)

**Comment by jbarlow83** [comment]
alternate works on OS X Preview and my iPhone. 

I did notice that spaces are sometimes missing in OS X's copy and paste text, while pdftotext shows the spaces, so perhaps it's not 100% but clearly this was the main issue.

> components of the relative motions of the fixed , stars with respect to the earth on the colour of thelightreachingusfromthem. Thelattereffect manifests itself in a slight displacement of the spectral lines of the light transmitted to us from

> a fixed star, as compared with the position of the same spectral lines when they are produced by a terrestrial source of light (Doppler principle). The experimental arguments in favour of the Maxwell-Lorentz theory, which are at the;same time arguments in favour of the theory of rela- tivity,aretoonumeroustobesetforthhere. In reality they limit the theoretical possibilities to such an extent, that no other theory than that of Maxwell and Lorentz has been able to hold its ownwhentestedbyexperience.

> But there are two classes of experimental facts hitherto obtained which can be represented in the Maxwell-Lorentz theory only by the introduction of an auxiliary hypothesis, which in itself—i.e. without making use of the theory of relativity— appears extraneous.

> Itisknownthatcathoderaysandtheso-called B—rays emitted by radioactive substances consist of negatively electrified particles (electrons) of verysmallinertiaandlargevelocity. By examin- ing the deflection of these rays under the influence of electric and magnetic fields, we can study the

> law of motion of these particles very exactly.

**Comment by behdad** [comment]
> Output:
> linn.pdf

For me, pdftotext outputs no text, but Evince, which also uses Poppler, correctly selects and extracts text.

**Comment by amitdo** [comment]
@behdad,

Try this:

> pdftotext linn.pdf -

**Comment by behdad** [comment]
> @behdad, try this:
> 
>     pdftotext linn.pdf -

Hah.  My bad.  Thanks :)

**Comment by jbreiden** [comment]
I got my hands on an iPad running iOS 9.2 and reproduced the problem. On iOS/Safari I cannot search 2.pdf (Ken Sharp's font) but can search with alternate.pdf (Behdad's font). Took me quite a while to figure out how how to make the search controls work.

So for your immediate problem, go ahead and substitute in Behdad's font into tessdata/pdf.ttf and you should be okay. We won't do that officially without a whole bunch more compatibility testing and reports, including the harder languages (Cherokee, vertical Japanese, Arabic) and additional renderers including Ghostscript and Firefox. Compatibility reports are appreciated.

Regarding the words running together on the Apple PDF renderer, that's not new. Apple PDF seems to do a worse job than everyone else at deciding word boundaries, and I've seen them screw up plenty of regular born-digital PDF files in the same way. Of course the root cause is the PDF spec itself, which does not explicitly define the concept of a word boundary. So I can't help you, but at least it isn't a regression. It's possible that Apple will get their act together a little better on this some day, but I have no reason to believe that it is on their radar.

**Comment by amitdo** [comment]
> There is also an alternative invisible font here, that contains an advanceWidth. I think it can be swapped in for tessdata/pdf.ttf. It has a side effect of making highlighting look even more bizarre in evince.

It looks terrible :(

**Comment by behdad** [comment]
My font has a huge advance width, because it was designed for another purpose.  Someone should create one with an advance width of 1024 instead of my 20480.

**Comment by jbreiden** [comment]
The PDF is keeping the advance width under control for Behdad's font. We're probably seeing something else. It's kind of cute zebra pattern. You get a black underline, and black boxes in all word gaps an in some letter gaps. (Obviously evince is doing a really bad  job, but this is much worse than with Ken Sharp's font, which highlights as a solid  black bar.) A little hard for me to investigate, since my copy of ttx is not cooperating.

![evince](https://cloud.githubusercontent.com/assets/4961958/12207036/566e60a8-b5f9-11e5-92c2-61ad905889b9.png)

**Comment by amitdo** [comment]
If you search for a phrase in evince, the highlighting looks more normal.
Strange!

**Comment by iikka-v** [comment]
FWIW, 
	1) I can confirm the problem as stated. Also, I've been using the same tesseract build, and it stopped working due to OS X update (unfortunately, I'm not sure which, possibly 10.11.1)
	2) As suggested above, using tofu.ttf fixed the issue for me (OS X 10.11.3, Finnish OCR), no recompile of tesseract needed

**Comment by jbreiden** [comment]
Partially blocked by https://github.com/behdad/fonttools/issues/497

**Comment by jbreiden** [comment]
Two choices. 
* Add advance width to Ken Sharp's font.
* Reduce advance with in Behdad's font.

This PDF is the former, please test for compatibility.

[simple-1.pdf](https://github.com/tesseract-ocr/tesseract/files/112801/simple-1.pdf)

```
diff -u pdf.ttx sharp.ttx
--- pdf.ttx	2016-02-01 10:24:02.875924041 -0800
+++ sharp.ttx	2016-02-01 10:23:38.659586076 -0800
@@ -14,7 +14,7 @@
     <checkSumAdjustment value="0xa737b34c"/>
     <magicNumber value="0x5f0f3cf5"/>
     <flags value="00000100 00000111"/>
-    <unitsPerEm value="256"/>
+    <unitsPerEm value="2048"/>
     <created value="Thu May 15 23:21:18 2014"/>
     <modified value="Thu May 15 23:21:18 2014"/>
     <xMin value="0"/>
@@ -33,7 +33,7 @@
     <ascent value="1"/>
     <descent value="-1"/>
     <lineGap value="0"/>
-    <advanceWidthMax value="0"/>
+    <advanceWidthMax value="1024"/>
     <minLeftSideBearing value="0"/>
     <minRightSideBearing value="0"/>
     <xMaxExtent value="0"/>
@@ -71,7 +71,7 @@
     <!-- The fields 'usFirstCharIndex' and 'usLastCharIndex'
          will be recalculated by the compiler -->
     <version value="3"/>
-    <xAvgCharWidth value="0"/>
+    <xAvgCharWidth value="1024"/>
     <usWeightClass value="400"/>
     <usWidthClass value="5"/>
     <fsType value="00000000 00000000"/>
@@ -122,7 +122,7 @@
 
   <hmtx>
     <mtx name=".notdef" width="0" lsb="0"/>
-    <mtx name=".null" width="0" lsb="0"/>
+    <mtx name=".null" width="1024" lsb="0"/>
   </hmtx>
 
   <cmap>
```

**Comment by iikka-v** [comment]
Nope, the PDF doen't seem to work for me (Mac OS X 10.11.3). The copied text is just equal number of spaces.

**Comment by jbreiden** [comment]
This PDF is the latter, please test for compatibility.

[behdad.pdf](https://github.com/tesseract-ocr/tesseract/files/112823/simple-1.pdf)

```
--- tofu.ttx	2016-02-01 10:17:15.038213397 -0800
+++ behdad.ttx	2016-02-01 10:43:29.839794297 -0800
@@ -33,7 +33,7 @@
     <ascent value="2048"/>
     <descent value="0"/>
     <lineGap value="0"/>
-    <advanceWidthMax value="20480"/>
+    <advanceWidthMax value="1024"/>
     <minLeftSideBearing value="0"/>
     <minRightSideBearing value="0"/>
     <xMaxExtent value="0"/>
@@ -69,7 +69,7 @@
 
   <OS_2>
     <version value="3"/>
-    <xAvgCharWidth value="790"/>
+    <xAvgCharWidth value="1024"/>
     <usWeightClass value="400"/>
     <usWidthClass value="5"/>
     <fsType value="00000000 00000000"/>
@@ -120,7 +120,7 @@
 
   <hmtx>
     <mtx name=".notdef" width="2048" lsb="0"/>
-    <mtx name="glyph00001" width="20480" lsb="0"/>
+    <mtx name="glyph00001" width="1024" lsb="0"/>
   </hmtx>
 
   <loca>
```

**Comment by iikka-v** [comment]
behdad.pdf works better. The letters are now reproduced correctly. There's still something funny with how selection works. Selecting from left to right doesn't correctly select all the letters. Right-to-left selects the three last characters one by one and then all four of the rest at once. Might be unrelated issue, though.

**Comment by jbreiden** [comment]
This is another attempt at the behdad font, with the contour data removed.
It fixes the visual problem with evince. Please test for compatibility. If it
works, we probably have a winner.

(Don't worry about the left-to-right selection stuff; that's due to mixing Hebrew
and English words in my test document)

[behdad2.pdf](https://github.com/tesseract-ocr/tesseract/files/112840/simple-1.pdf)

**Comment by iikka-v** [comment]
Nope, this does not work any more (selected characters are spaces again).

**Comment by jbreiden** [comment]
Utterly insane. I would really, really like to speak with the relevant software engineer at Apple.

**Comment by iikka-v** [comment]
Yes, utterly. I ran the tofu.ttf and the old pdf.ttf through Apples font validator. Both produced errors, but tofu.ttf only one, whereas the old pdf.ttf had additional "name table usability" errors. Please post the above font files (or diffs) and I'll run them through the validator as well. Perhaps this will give some insight to the issue.

**Comment by jbreiden** [comment]
Fonts as per request.

pdf.ttf - currently shipping font, by Ken Sharp
sharp.ttf - with advance width added

tofu.ttf - alternate font from behdad
behdad.ttf -  with advance with reduced
behdad2.ttf - with contour removed

[fonts.zip](https://github.com/tesseract-ocr/tesseract/files/112928/fonts.zip)

**Comment by iikka-v** [comment]
Thanks. Here's the verbose error report as given by Apples ftxvalidator (there's not really a version for 10.11, so some of this might be inaccurate). All report fatal errors and most errors are beyond my (admittedly limited) expertise on the subject. I hope they make more sense to you. 

TrueIVPro:fontTools.pkg iikka$ ./ftxvalidator -T -r -t all  /Users/iikka/Downloads/fonts-3/sharp/pdf.ttf 
pdf.ttf
	<no font name>
	A737B34C|
	kATSFontTestTrueTypeFontData
	Fatal Major Minor Technical Info 
	com.apple.TrueType.cmap.usability
		‘cmap’ table usability
		This test ensures the ‘cmap’ table has at least one usable subtable.
		NA
			(No messages)
	com.apple.TrueType.hmtx.usability
		‘hmtx’ table usability
		This test ensures the horizontal metrics table, ‘hmtx’, is valid.
		Minor 
			hmtx: Glyph 0 has a width of zero; since it is used for undisplayable characters, it should be non-zero.
				kATSFontTestSeverityMinorError
	com.apple.TrueType.loca.usability
		‘loca’ table usability
		This test ensures the glyph data location table, ‘loca’, is valid.
		NA
			(No messages)
	com.apple.TrueType.name.usability
		‘name’ table usability
		This test ensures the ‘name’ table is valid.
		Fatal 
			'name': The table has no PostScript name, which is required for printing and embedding in PDFs.
				kATSFontTestSeverityFatalError
			'name': The table has no full name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no family name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no style name, which is required
				kATSFontTestSeverityFatalError
	com.apple.TrueType.post.usability
		‘post’ table usability
		This test ensures the ‘post’ table is valid.
		Major 
			post: The table requires the font to have exactly 258 glyphs; it has 2
				kATSFontTestSeverityMajorError
	com.apple.TrueType.usability
		‘sfnt’ required tables
		This test ensures the ‘sfnt’ has all the required tables.
		Fatal 
			sfnt: The font has no PostScript name (required for printing and embedding in PDFs).
				kATSFontTestSeverityFatalError
	com.apple.TrueType.head.structure
		‘head’ table structure
		This test ensures the structure of the header table, ‘head’, accords with the specification.
		NA
			(No messages)
	com.apple.TrueType.hhea.structure
		‘hhea’ table structure
		This test ensures the structure of the horizontal header table, ‘hhea’, accords with the specification.
		Minor 
			hhea: ascent (1) is smaller than 40.00  of the em (256)
				kATSFontTestSeverityMinorError
			hhea: advanceMax is equal to zero
				kATSFontTestSeverityMinorError
	com.apple.TrueType.maxp.structure
		‘maxp’ table structure
		This test ensures the structure of the maximum profile table, ‘maxp’, accords with the specification.
		Minor 
			maxp: The maxZones field is equal to 1; it should be equal to 2
				kATSFontTestSeverityMinorError
	com.apple.TrueType.name.structure
		‘name’ table structure
		This test ensures the ‘name’ table is structurally sound.
		Fatal Minor 
			'name': A Unicode name at index 0 might be MacJapanese and not Unicode
				kATSFontTestSeverityMinorError
			'name': A Unicode name at index 0 might be MacKorean and not Unicode
				kATSFontTestSeverityMinorError
			'name': A Unicode name at index 0 might be MacSimpChinese and not Unicode
				kATSFontTestSeverityMinorError
			'name': A Unicode name at index 0 might be MacTradChinese and not Unicode
				kATSFontTestSeverityMinorError
			'name': Two name entries overlap:  name entry 0 (bytes 53-75) and name entry 2 (bytes 53-75)
				kATSFontTestSeverityMinorError
			'name': The table has no unique name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no PostScript name, which is required for printing and embedding in PDFs.
				kATSFontTestSeverityFatalError
			'name': The table has no family name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no style name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no full name, which is required
				kATSFontTestSeverityFatalError
	com.apple.TrueType.glyf.usability
		‘glyf’ table structure
		This test ensures the ‘glyf’ table is structurally sound.
		Minor Technical Info 
			glyf: Table has an odd number of bytes (1)
				kATSFontTestSeverityTechnicalError
			glyf: Glyph 0 has no visual appearance but should
				kATSFontTestSeverityMinorError
			glyf: Glyph 0 has no visual appearance
				kATSFontTestSeverityInformation
			glyf: Glyph 1 has no visual appearance
				kATSFontTestSeverityInformation
	com.apple.TrueType.glyf.instructions
		‘glyf’ table instructions
		This test ensures the instructions embedded in a ‘glyf’ table are sound.
		NA
			(No messages)
	com.apple.duplicates
		Duplicate fonts
		This test verifies the presence of installed fonts with the same name.
		NA
			(No messages)
TrueIVPro:fontTools.pkg iikka$ ./ftxvalidator -T -r -t all  /Users/iikka/Downloads/fonts-3/sharp/sharp.ttf 
sharp.ttf
	<no font name>
	FECBA532|
	kATSFontTestTrueTypeFontData
	Fatal Major Minor 
	com.apple.TrueType.cmap.usability
		‘cmap’ table usability
		This test ensures the ‘cmap’ table has at least one usable subtable.
		Fatal 
			cmap: The table is too small to hold enough data for the number of subtables it contains.
				kATSFontTestSeverityFatalError
	com.apple.TrueType.hmtx.usability
		‘hmtx’ table usability
		This test ensures the horizontal metrics table, ‘hmtx’, is valid.
		Minor 
			hmtx: Glyph 0 has a width of zero; since it is used for undisplayable characters, it should be non-zero.
				kATSFontTestSeverityMinorError
			hmtx: Glyph 1 has a width of 1024; it should be zero
				kATSFontTestSeverityMinorError
	com.apple.TrueType.loca.usability
		‘loca’ table usability
		This test ensures the glyph data location table, ‘loca’, is valid.
		NA
			(No messages)
	com.apple.TrueType.name.usability
		‘name’ table usability
		This test ensures the ‘name’ table is valid.
		Fatal 
			'name': The table has no PostScript name, which is required for printing and embedding in PDFs.
				kATSFontTestSeverityFatalError
			'name': The table has no full name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no family name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no style name, which is required
				kATSFontTestSeverityFatalError
	com.apple.TrueType.post.usability
		‘post’ table usability
		This test ensures the ‘post’ table is valid.
		Major 
			post: The table requires the font to have exactly 258 glyphs; it has 2
				kATSFontTestSeverityMajorError
	com.apple.TrueType.usability
		‘sfnt’ required tables
		This test ensures the ‘sfnt’ has all the required tables.
		Fatal Major 
			sfnt: The font contains 'sfnt' data but no recognizable glyph data.
				kATSFontTestSeverityFatalError
			sfnt: The font has some but not all of the tables required for TrueType outline data.
				kATSFontTestSeverityFatalError
			sfnt: The font has no PostScript name (required for printing and embedding in PDFs).
				kATSFontTestSeverityFatalError
			sfnt: Font table 2 ('glyf') has zero length.
				kATSFontTestSeverityMajorError
	com.apple.TrueType.head.structure
		‘head’ table structure
		This test ensures the structure of the header table, ‘head’, accords with the specification.
		NA
			(No messages)
	com.apple.TrueType.hhea.structure
		‘hhea’ table structure
		This test ensures the structure of the horizontal header table, ‘hhea’, accords with the specification.
		Minor 
			hhea: ascent (1) is smaller than 40.00  of the em (2048)
				kATSFontTestSeverityMinorError
	com.apple.TrueType.maxp.structure
		‘maxp’ table structure
		This test ensures the structure of the maximum profile table, ‘maxp’, accords with the specification.
		Minor 
			maxp: The maxZones field is equal to 1; it should be equal to 2
				kATSFontTestSeverityMinorError
	com.apple.TrueType.name.structure
		‘name’ table structure
		This test ensures the ‘name’ table is structurally sound.
		Fatal Minor 
			'name': A Unicode name at index 0 might be MacJapanese and not Unicode
				kATSFontTestSeverityMinorError
			'name': A Unicode name at index 0 might be MacKorean and not Unicode
				kATSFontTestSeverityMinorError
			'name': A Unicode name at index 0 might be MacSimpChinese and not Unicode
				kATSFontTestSeverityMinorError
			'name': A Unicode name at index 0 might be MacTradChinese and not Unicode
				kATSFontTestSeverityMinorError
			'name': Two name entries overlap:  name entry 0 (bytes 42-64) and name entry 2 (bytes 42-64)
				kATSFontTestSeverityMinorError
			'name': The table has no unique name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no PostScript name, which is required for printing and embedding in PDFs.
				kATSFontTestSeverityFatalError
			'name': The table has no family name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no style name, which is required
				kATSFontTestSeverityFatalError
			'name': The table has no full name, which is required
				kATSFontTestSeverityFatalError
	com.apple.duplicates
		Duplicate fonts
		This test verifies the presence of installed fonts with the same name.
		NA
			(No messages)
TrueIVPro:fontTools.pkg iikka$ ./ftxvalidator -T -r -t all  /Users/iikka/Downloads/fonts-3/behdad/behdad.ttf 
behdad.ttf
	<no font name>
	6D687F3C|
	kATSFontTestTrueTypeFontData
	Fatal Minor Info 
	com.apple.TrueType.cmap.usability
		‘cmap’ table usability
		This test ensures the ‘cmap’ table has at least one usable subtable.
		NA
			(No messages)
	com.apple.TrueType.hmtx.usability
		‘hmtx’ table usability
		This test ensures the horizontal metrics table, ‘hmtx’, is valid.
		Minor 
			hmtx: Glyph 1 has a width of 1024; it should be zero
				kATSFontTestSeverityMinorError
	com.apple.TrueType.loca.usability
		‘loca’ table usability
		This test ensures the glyph data location table, ‘loca’, is valid.
		NA
			(No messages)
	com.apple.TrueType.name.usability
		‘name’ table usability
		This test ensures the ‘name’ table is valid.
		NA
			(No messages)
	com.apple.TrueType.usability
		‘sfnt’ required tables
		This test ensures the ‘sfnt’ has all the required tables.
		Fatal 
			sfnt: The font is missing a name ('name') table (required by the TrueType specification).
				kATSFontTestSeverityFatalError
			sfnt: The font is missing a PostScript compatibility ('post') table (required by the TrueType specification).
				kATSFontTestSeverityFatalError
			sfnt: The font has no PostScript name (required for printing and embedding in PDFs).
				kATSFontTestSeverityFatalError
	com.apple.TrueType.head.structure
		‘head’ table structure
		This test ensures the structure of the header table, ‘head’, accords with the specification.
		NA
			(No messages)
	com.apple.TrueType.hhea.structure
		‘hhea’ table structure
		This test ensures the structure of the horizontal header table, ‘hhea’, accords with the specification.
		Minor 
			hhea: descent (0) is greater than or equal to zero
				kATSFontTestSeverityMinorError
	com.apple.TrueType.maxp.structure
		‘maxp’ table structure
		This test ensures the structure of the maximum profile table, ‘maxp’, accords with the specification.
		Minor 
			maxp: The maxZones field is equal to 1; it should be equal to 2
				kATSFontTestSeverityMinorError
	com.apple.TrueType.glyf.usability
		‘glyf’ table structure
		This test ensures the ‘glyf’ table is structurally sound.
		Minor Info 
			glyf: Glyph 0 has no visual appearance but should
				kATSFontTestSeverityMinorError
			glyf: Glyph 0 has no visual appearance
				kATSFontTestSeverityInformation
			glyf: Glyph 1 should have no visual appearance but does
				kATSFontTestSeverityMinorError
	com.apple.TrueType.glyf.instructions
		‘glyf’ table instructions
		This test ensures the instructions embedded in a ‘glyf’ table are sound.
		NA
			(No messages)
	com.apple.duplicates
		Duplicate fonts
		This test verifies the presence of installed fonts with the same name.
		NA
			(No messages)
TrueIVPro:fontTools.pkg iikka$ ./ftxvalidator -T -r -t all  /Users/iikka/Downloads/fonts-3/behdad/behdad2.ttf 
behdad2.ttf
	<no font name>
	9D684D32|
	kATSFontTestTrueTypeFontData
	Fatal Minor Info 
	com.apple.TrueType.cmap.usability
		‘cmap’ table usability
		This test ensures the ‘cmap’ table has at least one usable subtable.
		NA
			(No messages)
	com.apple.TrueType.hmtx.usability
		‘hmtx’ table usability
		This test ensures the horizontal metrics table, ‘hmtx’, is valid.
		Minor 
			hmtx: Glyph 1 has a width of 1024; it should be zero
				kATSFontTestSeverityMinorError
	com.apple.TrueType.loca.usability
		‘loca’ table usability
		This test ensures the glyph data location table, ‘loca’, is valid.
		NA
			(No messages)
	com.apple.TrueType.name.usability
		‘name’ table usability
		This test ensures the ‘name’ table is valid.
		NA
			(No messages)
	com.apple.TrueType.usability
		‘sfnt’ required tables
		This test ensures the ‘sfnt’ has all the required tables.
		Fatal 
			sfnt: The font is missing a name ('name') table (required by the TrueType specification).
				kATSFontTestSeverityFatalError
			sfnt: The font is missing a PostScript compatibility ('post') table (required by the TrueType specification).
				kATSFontTestSeverityFatalError
			sfnt: The font has no PostScript name (required for printing and embedding in PDFs).
				kATSFontTestSeverityFatalError
	com.apple.TrueType.head.structure
		‘head’ table structure
		This test ensures the structure of the header table, ‘head’, accords with the specification.
		NA
			(No messages)
	com.apple.TrueType.hhea.structure
		‘hhea’ table structure
		This test ensures the structure of the horizontal header table, ‘hhea’, accords with the specification.
		Minor 
			hhea: descent (0) is greater than or equal to zero
				kATSFontTestSeverityMinorError
	com.apple.TrueType.maxp.structure
		‘maxp’ table structure
		This test ensures the structure of the maximum profile table, ‘maxp’, accords with the specification.
		Minor 
			maxp: The maxZones field is equal to 1; it should be equal to 2
				kATSFontTestSeverityMinorError
	com.apple.TrueType.glyf.usability
		‘glyf’ table structure
		This test ensures the ‘glyf’ table is structurally sound.
		Minor Info 
			glyf: Glyph 0 has no visual appearance but should
				kATSFontTestSeverityMinorError
			glyf: Glyph 0 has no visual appearance
				kATSFontTestSeverityInformation
			glyf: Glyph 1 should have no visual appearance but does
				kATSFontTestSeverityMinorError
	com.apple.TrueType.glyf.instructions
		‘glyf’ table instructions
		This test ensures the instructions embedded in a ‘glyf’ table are sound.
		NA
			(No messages)
	com.apple.duplicates
		Duplicate fonts
		This test verifies the presence of installed fonts with the same name.
		NA
			(No messages)

**Comment by jbreiden** [comment]
Can you please make that report an attachment or something? The giant wall of text makes this bug harder to read.

**Comment by behdad** [comment]
> Partially blocked by behdad/fonttools#497

Fixed now.

**Comment by jbreiden** [comment]
For completeness, here is Ken Sharp's font with a contour added in. 

FONT
[sharp2.zip](https://github.com/tesseract-ocr/tesseract/files/119877/sharp2.zip)

PDF
[sharp2.pdf](https://github.com/tesseract-ocr/tesseract/files/119875/simple-1.pdf)

At this point, sharp2.ttf and behdad.ttf are the only fonts compatible with Apple Preview. They both come at the cost highlight aesthetics with evince. I think Preview is incorrect to require a contour for the glyph, and I think evince is incorrect to consider a contour when highlighting an invisible font. I do not have any reason so far to prefer one over the other, and I do not yet have compatibility test results from ghostscript, firefox, Microsoft Edge, etc.

**Comment by jbreiden** [comment]
I have filed a bug with Apple. This is not publicly visible and I do not know what the response will be. Noting it here simply simply for future reference. radr://24533090

**Comment by jbreiden** [comment]
In progress testing compatibility with candidates "sharp2" and "behdad" including getting some assistance with ghostscript. So far no user visible differences between them, and the former is the smaller change. Is there general consensus to work around the Apple compatibility problem, at the expense of Evince highlight aesthetics?

**Comment by jbarlow83** [comment]
@jbreiden I agree. OS X Preview is installed on ~10% of all desktop computers. Evince is just one of many PDF viewers for Linux users.

**Comment by bekirserifoglu** [comment]
@jbarlow83 and @jbreiden This bug also affects the Amazon Kindles. As an avid user of Amazon Kindle and Tesseract, I feel crippled now. And don't forget that all those pdfs generated with Tesseract won't work with Kindle either around the world.

**Comment by jbreiden** [comment]
@bekirserifoglu - can you please confirm that both proposed workarounds found in previous comments (sharp2.pdf, behdad.pdf) solve the problem on Kindle?

**Comment by bekirserifoglu** [comment]
@jbreiden I can confirm that both sharp and tofu fonts work great with Kindle Voyage and Preview on Os X. Feel free to mention me if you need anymore testing.

**Comment by jbreiden** [comment]
@bekirserifoglu - Is the failure case on Kindle broken search and broken copy-paste? Or is it even worse than that?

**Comment by bekirserifoglu** [comment]
@jbreiden Kindle just treats the pdf as a non-ocr'ed pdf. It is worse than OS X preview.

**Comment by jbreiden** [comment]
Okay, I've decided. We're going to use the sharp2 font.

For various embarassing reasons, I'd appreciate some help. Could someone please download this zip file, extract sharp2.ttf, and us it to replace pdf.ttf in the repository. The resulting file should still be called pdf.ttf. Again, I apologize for not doing this myself and promise to get my act together with respect to GitHub pull requests in the future.

[sharp2.zip](https://github.com/tesseract-ocr/tesseract/files/127261/sharp2.zip)

https://github.com/tesseract-ocr/tesseract/blob/master/tessdata/pdf.ttf

**Comment by jbarlow83** [comment]
Done. PR #220.

**Comment by iikka-v** [comment]
Thank you for fixing this.

**Comment by amitdo** [comment]
This was a short discussion... :-)
